File: src/main/java/june_2025/NumIslands.java
package june_2025;

public class NumIslands {

    public static void main(String[] args) {
        char[][] grid = {
                {'0', '1', '0', '1', '0'},
                {'1', '1', '0', '1', '0'},
                {'1', '1', '0', '0', '0'},
                {'0', '0', '0', '1', '1'}
        };

        NumIslands numIslands = new NumIslands();
        int result = numIslands.numIslands(grid);
        System.out.println(result);
    }

    // O(n*m) time | O(n*m) space
    public int numIslands(char[][] grid) {
        int count = 0;

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                if (grid[r][c] == '1') {
                    count++;
                    sink(r, c, grid);
                }
            }
        }

        return count;
    }

    private void sink(int r, int c, char[][] grid) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[r].length || grid[r][c] == '0') {
            return;
        }
        grid[r][c] = '0';
        sink(r + 1, c, grid);
        sink(r - 1, c, grid);
        sink(r, c + 1, grid);
        sink(r, c - 1, grid);
    }

}

File: python/num_islands.py
class Solution(object):
    # O(n * m) time | O(n * m) space
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count = 0

        def sink(r, c):
            if r < 0 or r >= len(grid) or c < 0 or c >= len(grid[r]) or grid[r][c] == "0":
                return
            grid[r][c] = "0"

            sink(r + 1, c)
            sink(r - 1, c)
            sink(r, c + 1)
            sink(r, c - 1)

        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == "1":
                    count += 1
                    sink(r, c)

        return count

grid = [
  ["1","1","0","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","1","1"]
]

solution = Solution()
print(solution.numIslands(grid))